All: Resolves #5244: Handles OneDrive throttling responses and sets User-Agent based on Microsoft best practices

alecjmaly opened the pull request:
<!--

Please prefix the title with the platform you are targetting:

Here are some examples of good titles:

- Desktop: Resolves #123: Added new setting to change font
- Mobile, Desktop: Fixes #456: Fixed config screen error
- All: Resolves #777: Made synchronisation faster

And here's an explanation of the title format:

- "Desktop" for the Windows/macOS/Linux app (Electron app)
- "Mobile" for the mobile app (or "Android" / "iOS" if the pull request only applies to one of the mobile platforms)
- "CLI" for the CLI app

If it's two platforms, separate them with commas - "Desktop, Mobile" or if it's for all platforms, prefix with "All".

If it's not related to any platform (such as a translation, change to the documentation, etc.), simply don't add a platform.

Then please append the issue that you've addressed or fixed. Use "Resolves #123" for new features or improvements and "Fixes #123" for bug fixes.

AND PLEASE READ THE GUIDE: https://github.com/laurent22/joplin/blob/dev/CONTRIBUTING.md

-->


# Fixes #5244 

## Note
I have created a new gulp task to generate the lib version number dynamically at build time, similar to the `compile-package-info.js` script in `/packages/app-desktop`. Please note that I have not kept the style of using a `/tools` directory and my added gulp task has been written inline within the `/packages/lib/gulpfile.js` file itself. @laurent22, Please let me know if you would like to redesign the gulp tasks to mirror the dte-desktop design (i.e. create a `/tools` within `packages/lib`) to keep code consistency, or if you prefer a different solution entirely. 

## References
- https://docs.microsoft.com/en-us/sharepoint/dev/general-development/how-to-avoid-getting-throttled-or-blocked-in-sharepoint-online#best-practices-to-handle-throttling

laurent22 commented on packages/lib/onedrive-api.ts:
```diff
@@ -265,11 +266,12 @@ export default class OneDriveApi {
 
 		for (let i = 0; i < 5; i++) {
 			options.headers['Authorization'] = `bearer ${this.token()}`;
+			options.headers['User-Agent'] = `ISV|Joplin|Joplin/${version}`;
```
I'm wondering here, should we put the lib version number or the app version number? It seems the app version would be more accurate? It can be retrieved using shim.appVersion().

Also I don't know the format of this user agent string, but shouldn't it be different for mobile/cli/desktop? Such as Joplin-Desktop, Joplin-Android, Joplin-iOS, Joplin-Cli (not sure about the format)? The platform is needed is in shim.platformName()

alecjmaly replied:
shim.appVersion() is great! I used the lib version as I didn't want to pull from another package and I honestly didn't look deep enough into shim to notice the functions being overloaded.

As for the user-agent, I'm just following the Microsoft Docs. The response returned is the same for all platforms so I don't think the user-agent being the same is an issue for OneDrive API requests. Unless you can think of a reason we should delineate between the apps?

**Source**
- https://docs.microsoft.com/en-us/sharepoint/dev/general-development/how-to-avoid-getting-throttled-or-blocked-in-sharepoint-online#how-to-decorate-your-http-traffic-to-avoid-throttling

![image](https://user-images.githubusercontent.com/32174181/127323629-12648796-a2b7-4a2e-96da-1cf3cb79e093.png)

The documentation is a bit strange, this article (above) is written for SharePoint but mentions OneDrive. Also, OneDrive is essentially a personal SharePoint site and operates under the [same api](https://docs.microsoft.com/en-us/onedrive/developer/rest-api/?view=odsp-graph-online). So we are trying to connect the dots and assume this change should help with the OneDrive throttling as well.

laurent22 replied:
Yes your user-agent is right then based on their documentation, it should be just "Joplin" as app name. And then let's used shim.appVersion for the version field then.

laurent22 commented on packages/lib/onedrive-api.ts:
```diff
@@ -265,11 +266,12 @@ export default class OneDriveApi {
+			options.headers['User-Agent'] = `ISV|Joplin|Joplin/${version}`;
 
-			const handleRequestRepeat = async (error: any) => {
+			const handleRequestRepeat = async (error: any, sleep_seconds: number = (i + 1) * 5) => {
```
It feels weird to have variable in the default parameter (I'm surprised it even works!). Could you perhaps set it to null instead, then have in the function `sleepSeconds = sleepSeconds === null ? (i + 1) * 5 : sleepSeconds`

Also please use camelCase for the variable names, so should be "sleepSeconds"

alecjmaly replied:
Have defaulted the parameter to null and implemented your suggestion using the null coalescing operator:
```javascript
sleepSeconds ??= (i + 1) * 5
```

laurent22 commented on packages/lib/onedrive-api.ts:
```diff
@@ -339,6 +341,15 @@ export default class OneDriveApi {
+					// Wait for OneDrive throttling
+					// Relavent Microsoft Docs: https://docs.microsoft.com/en-us/sharepoint/dev/general-development/how-to-avoid-getting-throttled-or-blocked-in-sharepoint-online#best-practices-to-handle-throttling
+					// Decrement retry count as multiple sync threads will cause repeated throttling errors - this will wait until throttling is resolved to continue, preventing a hard stop on the sync
+					i--;
```
Hmm, if OneDrive keeps sending the activityLimitReached will end up in an infinite loop, although I'm not sure what else we can do about it. We can't set an upper limit I guess since on initial sync there could be many of these throttle errors.

alecjmaly replied:
Indeed, I feel weird about this also. As I mentioned before, I assume given unlimited time/retries that the request will either succeed or result in a different error code which will then be handled normally. 

If you want to implement a failsafe we can try to think of an alternative. I tested with 20 threads and would get errors because they all would keep retrying but since there were multiple threads it kind of dragged the throttling longer for some threads which then exceeded the current 5 retry count and killed the sync completely.

laurent22 replied:
Not let's leave it as it is, and hopefully it won't be a problem. The fact that we're printing warnings to the log means we'll at least know what's happening.

laurent22 commented on packages/lib/onedrive-api.ts:
```diff
@@ -339,6 +341,15 @@ export default class OneDriveApi {
+					// Relavent Microsoft Docs: https://docs.microsoft.com/en-us/sharepoint/dev/general-development/how-to-avoid-getting-throttled-or-blocked-in-sharepoint-online#best-practices-to-handle-throttling
+					// Decrement retry count as multiple sync threads will cause repeated throttling errors - this will wait until throttling is resolved to continue, preventing a hard stop on the sync
+					i--;
+					const sleep_seconds = response.headers._headers['retry-after'][0];
```
sleepSeconds.

Could you also convert it to a number using `Number(response.headers._headers['retry-after'][0])` and perhaps also validate with `isNaN` just to be sure. If it's NaN (very unlikely), maybe just throw an error.

alecjmaly replied:
Converted to Number when passing to function and test for isNaN() in if statement

```javascript
else if (error?.code === 'activityLimitReached' && response?.headers?._headers['retry-after'][0] && !isNaN(Number(response?.headers?._headers['retry-after'][0]))) {
  // Wait for OneDrive throttling
  // Relavent Microsoft Docs: https://docs.microsoft.com/en-us/sharepoint/dev/general-development/how-to-avoid-getting-throttled-or-blocked-in-sharepoint-online#best-practices-to-handle-throttling
  // Decrement retry count as multiple sync threads will cause repeated throttling errors - this will wait until throttling is resolved to continue, preventing a hard stop on the sync
  i--;
  const sleepSeconds = response.headers._headers['retry-after'][0];
  logger.info(`OneDrive Throttle, sync thread sleeping for ${sleepSeconds} seconds...`);
  await handleRequestRepeat(error, Number(sleepSeconds));
  continue;
  }
```

laurent22 commented on packages/lib/onedrive-api.ts:
```diff
@@ -339,6 +341,15 @@ export default class OneDriveApi {
+					i--;
+					const sleep_seconds = response.headers._headers['retry-after'][0];
+					logger.info(`OneDrive Throttle, sync thread sleeping for ${sleep_seconds} seconds...`);
+					await handleRequestRepeat(error, parseInt(sleep_seconds));
```
parseInt is kind of broken because if you don't provide a radix it will try to guess and sometimes does it incorrectly. I think it's best to stick to `Number(string)`

laurent22 commented on packages/lib/onedrive-api.ts:
```diff
@@ -339,6 +341,15 @@ export default class OneDriveApi {
+					// Decrement retry count as multiple sync threads will cause repeated throttling errors - this will wait until throttling is resolved to continue, preventing a hard stop on the sync
+					i--;
+					const sleep_seconds = response.headers._headers['retry-after'][0];
+					logger.info(`OneDrive Throttle, sync thread sleeping for ${sleep_seconds} seconds...`);
```
Makes sense to at least log it. Do you know how long they make us wait in general? As you mentioned if it takes a very long time perhaps we should display this to the user, although I'm not sure how at the moment.

alecjmaly replied:
I tested with a very aggressive 20 "max concurrent connections". 
It took about 15min before the bulk download could resume. 

The typical 'retry-count' header maxes around 5min, but with 20 threads it seemed to drag on. Essentially, you end up with 20 threads retrying every `retry-count HTTP header` seconds. If it fails due to throttling, that request still counts against your request quota and increases the duration of throttling. 

Less threads will not have such large wait times, I would imagine.

Unfortunately, as far as I can tell, Microsoft doesn't explicitly say what the limits are, so it's hard to say for certian.

As pointed out by another contributor, the batch endpoint would be a great enhancement, however, it would take a bit of a refactor.

laurent22 reviewed (commented):
Thanks for the update. Overall it looks good.

Could you also check that the test units are working with your changes please? You would need to set the sync target to "onedrive" there:

https://github.com/laurent22/joplin/blob/6d86db3cb09e6e5f41769df4b7c19ead04a498ea/packages/lib/testing/test-utils.ts#L138

Then here it's explained how to set it up for OneDrive:

https://github.com/laurent22/joplin/blob/6d86db3cb09e6e5f41769df4b7c19ead04a498ea/packages/lib/testing/test-utils.ts#L537

The credential file should go under `~/joplin-credentials/`

Then run the tests from packages/lib using `npm test -- --runInBand`

laurent22 commented:
Also see CI for the linter errors.

alecjmaly commented:
I have resolved linter issues, however, some tests are failing.

![image](https://user-images.githubusercontent.com/32174181/127338142-791d6650-5cef-4cfd-8688-8478ac5cf094.png)


They seem to fail from the dev branch without modifications as well, I don't think any of my changes are causing these tests to fail. Do you know anything about these? 
The other failed tests are similar `services_SearchFilter > search type nonlatin` tests

alecjmaly commented:
Linting errors should be resolved, implemented your suggestions and cleaned up my unused changes.
The main issue I see now is a few failing tests, which seem to fail from a clean build as well.

![image](https://user-images.githubusercontent.com/32174181/127340937-06274cf2-1287-4046-93f7-6941d6f3b04d.png)

Do these tests fail for you as well? 
All failing tests are `services_SearchFilter > search type nonlatin`

The tests seem to fail on the `dev` branch as well, which I used as my base. 
Would you prefer me to base my changes on another branch?

laurent22 commented:
That's strange, how are you running the tests?

alecjmaly commented:
> That's strange, how are you running the tests?

As you mentioned before.
`npm test -- --runInBand`

alecjmaly commented:
Definitely strange that your builds are passing (or that mine are failing).

I literally just pulled from the dev branch on the Joplin repo and it fails.

![image](https://user-images.githubusercontent.com/32174181/127350211-9f52e6b1-406f-4643-a4e2-327343290faf.png)

```bash
date
git remote get-url origin
git branch
git pull
./node_modules/.bin/jest ./services/searchengine/SearchFilter.test.js -t 'should support filtering by updated date' --verbose false
``` 

I will spend some time trying to see if I can understand what's going on.
I just wanted to provide an update with proofs of what I'm seeing. 

Very odd.

alecjmaly commented:
Ok, after some debugging I believe this is related to timezones.
I am in Eastern Standard Time (EST)

## Test code `packages/lib/services/searchengine/SearchFilter.test.js`
The Jest testing code inserts timestamps into the database by `Date.parse()`
Based on the [docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/parse), this enters the record using UTC/GMT time.

![image](https://user-images.githubusercontent.com/32174181/127364838-522a80ed-95bf-4f2e-a6d5-552219c4d2d0.png)

### Results from entering records into database

![image](https://user-images.githubusercontent.com/32174181/127365128-29425904-91c9-4228-b26c-fc6627bc0148.png)

```text
{
    title: 'I updated this on',
    body: 'May 20 2020',
    updated_time: 1589932800000,
    user_updated_time: 1589932800000,
    source: 'joplin',
    source_application: 'net.cozic.joplintest-cli',
    order: 1627489647402,
    id: 'fa82bbf3ee484623a82ced0fbfda7e68',
    created_time: 1627489647402,
    user_created_time: 1627489647402,
    type_: 1
}
```


### `engine.search()`
![image](https://user-images.githubusercontent.com/32174181/127365436-bc686a1a-6bd8-4256-beed-e7b167e34706.png)

Note that engine.search() will eventually build a sql query from `formatLocalToMs()` in `packages/lib/time.ts` using `moment.js`

![image](https://user-images.githubusercontent.com/32174181/127366067-ee994095-ac14-4416-a3a8-c1d9a7db8dab.png)

Note that moment.js [defaults to local time](https://momentjs.com/timezone/docs/#/using-timezones/default-timezone/).

![image](https://user-images.githubusercontent.com/32174181/127366443-1ff31053-57d0-4d2a-95ab-b3ed569f8bbb.png)


This results in the following query:

```text
query = "
    SELECT
    notes.id,
    notes.title,

    notes.user_created_time,
    notes.user_updated_time,
    notes.is_todo,
    notes.todo_completed,
    notes.parent_id
    FROM notes
    WHERE 1 
    AND (  ROWID IN (
        SELECT ROWID
        FROM notes
        WHERE notes.user_updated_time >= ?
    ))
"


params = "1589947200000"
```

# Timestamp Differences

## real datetime (inserted by `Date.parse()` (utc/gmt):
1589932800000

![image](https://user-images.githubusercontent.com/32174181/127364594-e0889f0e-2379-4eab-bad3-b2a44cfa0e20.png)

## query generated from moment.js (EST - in my case):
1589947200000

![image](https://user-images.githubusercontent.com/32174181/127366858-46829e84-c14f-4a11-ba04-002291aac5b7.png)

## Result
Since the following is **FALSE** `(inserted timestamp) 1589932800000 >= (search query timestamp) 1589947200000` - no records are returned.

# Testing the theory

If I were to set the `Date.parse()` function to use my local timestamp, the tests succeed.

![image](https://user-images.githubusercontent.com/32174181/127366982-b3766c58-32cf-4da8-843b-068c493606ce.png)

![image](https://user-images.githubusercontent.com/32174181/127367018-66b33f35-f653-41a3-9130-0b335da73792.png)


# Conclusions / Discussion

So, I'm curious, what timezone are you in?
I believe this is why my local tests fail, however, this does not explain the github CI pipeline. I would guess github is spinning up a VM somewhere in my timezone and somewhere else for yours? This is strange.

What are your thoughts on this?
I'm surprised this hasn't come up with other contributors, perhaps I'm mis-assessing the situation?

alecjmaly commented:
Not sure why the checks passed now but ok...

tessus commented:
Good find. These tests always fail for me too locally. I'm also in EST. Apparently the gh runners use UTC by default.

alecjmaly commented:
> Good find. These tests always fail for me too locally. I'm also in EST. Apparently the gh runners use UTC by default.

Yup, that makes sense. I believe my last checks failed because I accidently committed back the onedrive tests as uncommented - the absence of a credential file lead to a failure, I bet.

As far as I can tell I'm done with updates on this, feel free to review at your leisure. Thanks for all your input thus far!

leematos commented:
@alecjmaly nice find on those test failures! I was seeing the same in dev on my local in EST. <3

laurent22 commented:
Ok looks like we're good now, thanks for the fix @alecjmaly!
